semgrep/semgrep/core_runner.py

```diff
@@ -199,99 +199,98 @@ def _run_rule(
+            patterns_regex, patterns = partition(
+                lambda p: p.expression.operator == OPERATORS.REGEX,
+                all_patterns_for_language,
+            )
+            if patterns_regex:
+                patterns_json = [pattern.to_json() for pattern in patterns_regex]
 
-            for language, all_patterns_for_language in self._group_patterns_by_language(
-                [rule]
-            ).items():
-                targets = target_manager.get_files(
-                    language, rule.includes, rule.excludes
-                )
-                if targets == []:
-                    continue
+                try:
+                    patterns_re = [
+                        (pattern["id"], re.compile(pattern["pattern"]))
+                        for pattern in patterns_json
+                    ]
+                except re.error as err:
+                    raise SemgrepError(f"invalid regular expression specified: {err}")
 
-                # semgrep-core doesn't know about OPERATORS.REGEX - this is
-                # strictly a semgrep Python feature. Regex filtering is
-                # performed purely in Python code then compared against
-                # semgrep-core's results for other patterns.
-                patterns_regex, patterns = partition(
-                    lambda p: p.expression.operator == OPERATORS.REGEX,
-                    all_patterns_for_language,
+                re_fn = functools.partial(get_re_matches, patterns_re)
+                with multiprocessing.Pool(self._jobs) as pool:
+                    matches = pool.map(re_fn, targets)
+
+                outputs.extend(
+                    single_match
+                    for file_matches in matches
+                    for single_match in file_matches
                 )
-                if patterns_regex:
-                    patterns_json = [pattern.to_json() for pattern in patterns_regex]
 
-                    try:
-                        patterns_re = [
-                            (pattern["id"], re.compile(pattern["pattern"]))
-                            for pattern in patterns_json
-                        ]
-                    except re.error as err:
-                        raise SemgrepError(
-                            f"invalid regular expression specified: {err}"
-                        )
+            patterns_json = [p.to_json() for p in patterns]
+            # very important not to sort keys here
+            with tempfile.NamedTemporaryFile(
+                "w"
+            ) as pattern_file, tempfile.NamedTemporaryFile(
+                "w"
+            ) as target_file, tempfile.NamedTemporaryFile(
+                "w"
+            ) as equiv_file:
+                yaml = YAML()
+                yaml.dump({"rules": patterns_json}, pattern_file)
+                pattern_file.flush()
+                target_file.write("\n".join([str(t) for t in targets]))
```

Comment:
We can do `"\n".join(str(t) for t in targets)` (no list comp) to be a bit more memory efficient.